### ISSUE TITLE
[Bug] [Kotlin] ReachingDefPass failing for sample repo

### ISSUE THREAD

**Issue by tuxology** [joernio/joern#3706]
**Describe the bug**
Pass failing for sample repo: https://github.com/shockbytes/Corey 

**To Reproduce**
`importCode("Corey")`

**Expected behavior**
CPG gets generated

**Error Log**
```
2023-10-01 17:15:09.184 ERROR CpgPassBase: Pass io.joern.dataflowengineoss.passes.reachingdef.ReachingDefPass failed
java.util.NoSuchElementException: java.util.NoSuchElementException: key not found: io.shiftleft.codepropertygraph.generated.nodes.MethodReturn[label=METHOD_RETURN; id=43908]
        at jdk.internal.reflect.DirectConstructorHandleAccessor.newInstance(DirectConstructorHandleAccessor.java:67) ~[?:?]
        at java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:499) ~[?:?]
        at java.lang.reflect.Constructor.newInstance(Constructor.java:483) ~[?:?]
        at java.util.concurrent.ForkJoinTask.getThrowableException(ForkJoinTask.java:562) ~[?:?]
        at java.util.concurrent.ForkJoinTask.reportException(ForkJoinTask.java:591) ~[?:?]
        at java.util.concurrent.ForkJoinTask.invoke(ForkJoinTask.java:689) ~[?:?]
        at java.util.stream.ReduceOps$ReduceOp.evaluateParallel(ReduceOps.java:927) ~[?:?]
        at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:233) ~[?:?]
        at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:693) ~[?:?]
        at io.shiftleft.passes.NewStyleCpgPassBase.runWithBuilder(CpgPass.scala:152) ~[io.shiftleft.codepropertygraph_3-1.4.22.jar:1.4.22]
        at io.shiftleft.passes.ForkJoinParallelCpgPass.createApplySerializeAndStore(CpgPass.scala:74) ~[io.shiftleft.codepropertygraph_3-1.4.22.jar:1.4.22]
        at io.shiftleft.semanticcpg.layers.LayerCreator.runPass(LayerCreator.scala:53) ~[io.joern.semanticcpg-2.0.103.jar:2.0.103]
        at io.joern.dataflowengineoss.layers.dataflows.OssDataFlow.create$$anonfun$1(OssDataFlow.scala:31) ~[io.joern.dataflowengineoss-2.0.103.jar:2.0.103]
        at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15) ~[org.scala-lang.scala3-library_3-3.3.1.jar:3.3.1]
        at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10) ~[org.scala-lang.scala3-library_3-3.3.1.jar:3.3.1]
        at scala.collection.IterableOnceOps.foreach(IterableOnce.scala:575) ~[org.scala-lang.scala-library-2.13.10.jar:?]
        at scala.collection.IterableOnceOps.foreach$(IterableOnce.scala:573) ~[org.scala-lang.scala-library-2.13.10.jar:?]
        at scala.collection.AbstractIterator.foreach(Iterator.scala:1300) ~[org.scala-lang.scala-library-2.13.10.jar:?]
        at io.joern.dataflowengineoss.layers.dataflows.OssDataFlow.create(OssDataFlow.scala:32) ~[io.joern.dataflowengineoss-2.0.103.jar:2.0.103]
        at io.shiftleft.semanticcpg.layers.LayerCreator.run(LayerCreator.scala:32) ~[io.joern.semanticcpg-2.0.103.jar:2.0.103]
        at io.joern.console.Console.runCreator(Console.scala:443) ~[io.joern.console-2.0.103.jar:2.0.103]
        at io.joern.console.Console._runAnalyzer$$anonfun$1(Console.scala:432) ~[io.joern.console-2.0.103.jar:2.0.103]
        at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15) ~[org.scala-lang.scala3-library_3-3.3.1.jar:3.3.1]
        at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10) ~[org.scala-lang.scala3-library_3-3.3.1.jar:3.3.1]
        at scala.collection.IterableOnceOps.foreach(IterableOnce.scala:575) ~[org.scala-lang.scala-library-2.13.10.jar:?]
        at scala.collection.IterableOnceOps.foreach$(IterableOnce.scala:573) ~[org.scala-lang.scala-library-2.13.10.jar:?]
        at scala.collection.AbstractIterable.foreach(Iterable.scala:933) ~[org.scala-lang.scala-library-2.13.10.jar:?]
        at io.joern.console.Console._runAnalyzer(Console.scala:433) ~[io.joern.console-2.0.103.jar:2.0.103]
        at io.joern.joerncli.console.JoernConsole.applyDefaultOverlays(JoernConsole.scala:44) ~[io.joern.joern-cli-2.0.103.jar:2.0.103]
        at io.joern.console.cpgcreation.ImportCode.apply$$anonfun$1(ImportCode.scala:147) ~[io.joern.console-2.0.103.jar:2.0.103]
        at scala.Option.map(Option.scala:242) ~[org.scala-lang.scala-library-2.13.10.jar:?]
        at io.joern.console.cpgcreation.ImportCode.io$joern$console$cpgcreation$ImportCode$$apply(ImportCode.scala:148) ~[io.joern.console-2.0.103.jar:2.0.103]
        at io.joern.console.cpgcreation.ImportCode.apply(ImportCode.scala:40) ~[io.joern.console-2.0.103.jar:2.0.103]
        at rs$line$2$.<clinit>(rs$line$2:1) ~[?:?]
        at rs$line$2.res0(rs$line$2) ~[?:?]
        at jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104) ~[?:?]
        at java.lang.reflect.Method.invoke(Method.java:577) ~[?:?]
        at replpp.Rendering.$anonfun$4(Rendering.scala:98) ~[com.michaelpollmeier.scala-repl-pp_3-0.1.70.jar:0.1.70]
        at scala.Option.flatMap(Option.scala:283) ~[org.scala-lang.scala-library-2.13.10.jar:?]
        at replpp.Rendering.valueOf(Rendering.scala:99) ~[com.michaelpollmeier.scala-repl-pp_3-0.1.70.jar:0.1.70]
        at replpp.Rendering.renderVal(Rendering.scala:142) ~[com.michaelpollmeier.scala-repl-pp_3-0.1.70.jar:0.1.70]
        at replpp.DottyReplDriver.$anonfun$7(DottyReplDriver.scala:364) ~[com.michaelpollmeier.scala-repl-pp_3-0.1.70.jar:0.1.70]
        at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15) ~[org.scala-lang.scala3-library_3-3.3.1.jar:3.3.1]
        at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10) ~[org.scala-lang.scala3-library_3-3.3.1.jar:3.3.1]
        at scala.collection.immutable.List.foreach(List.scala:333) ~[org.scala-lang.scala-library-2.13.10.jar:?]
        at replpp.DottyReplDriver.extractAndFormatMembers$1(DottyReplDriver.scala:370) ~[com.michaelpollmeier.scala-repl-pp_3-0.1.70.jar:0.1.70]
        at replpp.DottyReplDriver.renderDefinitions$$anonfun$2(DottyReplDriver.scala:402) ~[com.michaelpollmeier.scala-repl-pp_3-0.1.70.jar:0.1.70]
        at scala.Option.map(Option.scala:242) ~[org.scala-lang.scala-library-2.13.10.jar:?]
        at replpp.DottyReplDriver.renderDefinitions(DottyReplDriver.scala:407) ~[com.michaelpollmeier.scala-repl-pp_3-0.1.70.jar:0.1.70]
        at replpp.DottyReplDriver.compile$$anonfun$2(DottyReplDriver.scala:304) ~[com.michaelpollmeier.scala-repl-pp_3-0.1.70.jar:0.1.70]
        at scala.util.Either.fold(Either.scala:189) ~[org.scala-lang.scala-library-2.13.10.jar:?]
        at replpp.DottyReplDriver.compile(DottyReplDriver.scala:322) ~[com.michaelpollmeier.scala-repl-pp_3-0.1.70.jar:0.1.70]
        at replpp.DottyReplDriver.interpret(DottyReplDriver.scala:245) ~[com.michaelpollmeier.scala-repl-pp_3-0.1.70.jar:0.1.70]
        at replpp.ReplDriverBase.interpretInput(ReplDriverBase.scala:65) ~[com.michaelpollmeier.scala-repl-pp_3-0.1.70.jar:0.1.70]
        at replpp.ReplDriver.loop$1$$anonfun$1(ReplDriver.scala:35) ~[com.michaelpollmeier.scala-repl-pp_3-0.1.70.jar:0.1.70]
        at scala.util.Try$.apply(Try.scala:210) ~[org.scala-lang.scala-library-2.13.10.jar:?]
        at replpp.ReplDriver.loop$1(ReplDriver.scala:36) ~[com.michaelpollmeier.scala-repl-pp_3-0.1.70.jar:0.1.70]
        at replpp.ReplDriver.runUntilQuit$$anonfun$2(ReplDriver.scala:52) ~[com.michaelpollmeier.scala-repl-pp_3-0.1.70.jar:0.1.70]
        at replpp.DottyReplDriver.withRedirectedOutput(DottyReplDriver.scala:187) ~[com.michaelpollmeier.scala-repl-pp_3-0.1.70.jar:0.1.70]
        at replpp.DottyReplDriver.runBody$$anonfun$1(DottyReplDriver.scala:161) ~[com.michaelpollmeier.scala-repl-pp_3-0.1.70.jar:0.1.70]
        at dotty.tools.runner.ScalaClassLoader$.asContext(ScalaClassLoader.scala:80) [org.scala-lang.scala3-compiler_3-3.3.1.jar:3.3.1]
        at replpp.DottyReplDriver.runBody(DottyReplDriver.scala:161) [com.michaelpollmeier.scala-repl-pp_3-0.1.70.jar:0.1.70]
        at replpp.ReplDriver.runUntilQuit(ReplDriver.scala:53) [com.michaelpollmeier.scala-repl-pp_3-0.1.70.jar:0.1.70]
        at replpp.InteractiveShell$.run(InteractiveShell.scala:39) [com.michaelpollmeier.scala-repl-pp_3-0.1.70.jar:0.1.70]
        at io.joern.console.InteractiveShell.startInteractiveShell(BridgeBase.scala:237) [io.joern.console-2.0.103.jar:2.0.103]
        at io.joern.console.InteractiveShell.startInteractiveShell$(BridgeBase.scala:224) [io.joern.console-2.0.103.jar:2.0.103]
        at io.joern.joerncli.console.ReplBridge$.startInteractiveShell(ReplBridge.scala:7) [io.joern.joern-cli-2.0.103.jar:2.0.103]
        at io.joern.console.BridgeBase.run(BridgeBase.scala:204) [io.joern.console-2.0.103.jar:2.0.103]
        at io.joern.console.BridgeBase.run$(BridgeBase.scala:44) [io.joern.console-2.0.103.jar:2.0.103]
        at io.joern.joerncli.console.ReplBridge$.run(ReplBridge.scala:7) [io.joern.joern-cli-2.0.103.jar:2.0.103]
        at io.joern.joerncli.console.ReplBridge$.main(ReplBridge.scala:12) [io.joern.joern-cli-2.0.103.jar:2.0.103]
        at io.joern.joerncli.console.ReplBridge.main(ReplBridge.scala) [io.joern.joern-cli-2.0.103.jar:2.0.103]
Caused by: java.util.NoSuchElementException: key not found: io.shiftleft.codepropertygraph.generated.nodes.MethodReturn[label=METHOD_RETURN; id=43908]
        at scala.collection.immutable.BitmapIndexedMapNode.apply(HashMap.scala:631) ~[org.scala-lang.scala-library-2.13.10.jar:?]
        at scala.collection.immutable.HashMap.apply(HashMap.scala:132) ~[org.scala-lang.scala-library-2.13.10.jar:?]
        at io.joern.dataflowengineoss.passes.reachingdef.ReachingDefFlowGraph.pred(ReachingDefProblem.scala:72) ~[io.joern.dataflowengineoss-2.0.103.jar:2.0.103]
        at io.joern.dataflowengineoss.passes.reachingdef.ReachingDefFlowGraph.pred(ReachingDefProblem.scala:71) ~[io.joern.dataflowengineoss-2.0.103.jar:2.0.103]
        at io.joern.dataflowengineoss.passes.reachingdef.DataFlowSolver.$anonfun$1(DataFlowSolver.scala:20) ~[io.joern.dataflowengineoss-2.0.103.jar:2.0.103]
        at scala.collection.StrictOptimizedIterableOps.flatMap(StrictOptimizedIterableOps.scala:118) ~[org.scala-lang.scala-library-2.13.10.jar:?]
        at scala.collection.StrictOptimizedIterableOps.flatMap$(StrictOptimizedIterableOps.scala:105) ~[org.scala-lang.scala-library-2.13.10.jar:?]
        at scala.collection.mutable.ListBuffer.flatMap(ListBuffer.scala:39) ~[org.scala-lang.scala-library-2.13.10.jar:?]
        at io.joern.dataflowengineoss.passes.reachingdef.DataFlowSolver.calculateMopSolutionForwards(DataFlowSolver.scala:33) ~[io.joern.dataflowengineoss-2.0.103.jar:2.0.103]
        at io.joern.dataflowengineoss.passes.reachingdef.ReachingDefPass.runOnPart(ReachingDefPass.scala:31) ~[io.joern.dataflowengineoss-2.0.103.jar:2.0.103]
        at io.joern.dataflowengineoss.passes.reachingdef.ReachingDefPass.runOnPart(ReachingDefPass.scala:23) ~[io.joern.dataflowengineoss-2.0.103.jar:2.0.103]
        at io.shiftleft.passes.NewStyleCpgPassBase$$anon$2.accept(CpgPass.scala:147) ~[io.shiftleft.codepropertygraph_3-1.4.22.jar:1.4.22]
        at io.shiftleft.passes.NewStyleCpgPassBase$$anon$2.accept(CpgPass.scala:146) ~[io.shiftleft.codepropertygraph_3-1.4.22.jar:1.4.22]
        at java.util.stream.ReduceOps$4ReducingSink.accept(ReduceOps.java:220) ~[?:?]
        at java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:992) ~[?:?]
        at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509) ~[?:?]
        at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499) ~[?:?]
        at java.util.stream.ReduceOps$ReduceTask.doLeaf(ReduceOps.java:960) ~[?:?]
        at java.util.stream.ReduceOps$ReduceTask.doLeaf(ReduceOps.java:934) ~[?:?]
        at java.util.stream.AbstractTask.compute(AbstractTask.java:327) ~[?:?]
        at java.util.concurrent.CountedCompleter.exec(CountedCompleter.java:754) ~[?:?]
        at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373) ~[?:?]
        at java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1182) ~[?:?]
        at java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1655) ~[?:?]
        at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1622) ~[?:?]
        at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165) ~[?:?]
java.util.NoSuchElementException: java.util.NoSuchElementException: key not found: io.shiftleft.codepropertygraph.generated.nodes.MethodReturn[label=METHOD_RETURN; id=43908]
  at java.base/jdk.internal.reflect.DirectConstructorHandleAccessor.newInstance(DirectConstructorHandleAccessor.java:67)
  at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:499)
  at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:483)
  at java.base/java.util.concurrent.ForkJoinTask.getThrowableException(ForkJoinTask.java:562)
  at java.base/java.util.concurrent.ForkJoinTask.reportException(ForkJoinTask.java:591)
  at java.base/java.util.concurrent.ForkJoinTask.invoke(ForkJoinTask.java:689)
  at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateParallel(ReduceOps.java:927)
  at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:233)
  at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:693)
  at io.shiftleft.passes.NewStyleCpgPassBase.runWithBuilder(CpgPass.scala:152)
  at io.shiftleft.passes.ForkJoinParallelCpgPass.createApplySerializeAndStore(CpgPass.scala:74)
  at io.shiftleft.semanticcpg.layers.LayerCreator.runPass(LayerCreator.scala:53)
  at io.joern.dataflowengineoss.layers.dataflows.OssDataFlow.create$$anonfun$1(OssDataFlow.scala:31)
  at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
  at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
  at scala.collection.IterableOnceOps.foreach(IterableOnce.scala:575)
  at scala.collection.IterableOnceOps.foreach$(IterableOnce.scala:573)
  at scala.collection.AbstractIterator.foreach(Iterator.scala:1300)
  at io.joern.dataflowengineoss.layers.dataflows.OssDataFlow.create(OssDataFlow.scala:32)
  at io.shiftleft.semanticcpg.layers.LayerCreator.run(LayerCreator.scala:32)
  at io.joern.console.Console.runCreator(Console.scala:443)
  at io.joern.console.Console._runAnalyzer$$anonfun$1(Console.scala:432)
  at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
  at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
  at scala.collection.IterableOnceOps.foreach(IterableOnce.scala:575)
  at scala.collection.IterableOnceOps.foreach$(IterableOnce.scala:573)
  at scala.collection.AbstractIterable.foreach(Iterable.scala:933)
  at io.joern.console.Console._runAnalyzer(Console.scala:433)
  at io.joern.joerncli.console.JoernConsole.applyDefaultOverlays(JoernConsole.scala:44)
  at io.joern.console.cpgcreation.ImportCode.apply$$anonfun$1(ImportCode.scala:147)
  at scala.Option.map(Option.scala:242)
  at io.joern.console.cpgcreation.ImportCode.io$joern$console$cpgcreation$ImportCode$$apply(ImportCode.scala:148)
  at io.joern.console.cpgcreation.ImportCode.apply(ImportCode.scala:40)
  ... 39 elided
Caused by: java.util.NoSuchElementException: key not found: io.shiftleft.codepropertygraph.generated.nodes.MethodReturn[label=METHOD_RETURN; id=43908]
  at scala.collection.immutable.BitmapIndexedMapNode.apply(HashMap.scala:631)
  at scala.collection.immutable.HashMap.apply(HashMap.scala:132)
  at io.joern.dataflowengineoss.passes.reachingdef.ReachingDefFlowGraph.pred(ReachingDefProblem.scala:72)
  at io.joern.dataflowengineoss.passes.reachingdef.ReachingDefFlowGraph.pred(ReachingDefProblem.scala:71)
  at io.joern.dataflowengineoss.passes.reachingdef.DataFlowSolver.$anonfun$1(DataFlowSolver.scala:20)
  at scala.collection.StrictOptimizedIterableOps.flatMap(StrictOptimizedIterableOps.scala:118)
  at scala.collection.StrictOptimizedIterableOps.flatMap$(StrictOptimizedIterableOps.scala:105)
  at scala.collection.mutable.ListBuffer.flatMap(ListBuffer.scala:39)
  at io.joern.dataflowengineoss.passes.reachingdef.DataFlowSolver.calculateMopSolutionForwards(DataFlowSolver.scala:33)
  at io.joern.dataflowengineoss.passes.reachingdef.ReachingDefPass.runOnPart(ReachingDefPass.scala:31)
  at io.joern.dataflowengineoss.passes.reachingdef.ReachingDefPass.runOnPart(ReachingDefPass.scala:23)
  at io.shiftleft.passes.NewStyleCpgPassBase$$anon$2.accept(CpgPass.scala:147)
  at io.shiftleft.passes.NewStyleCpgPassBase$$anon$2.accept(CpgPass.scala:146)
  at java.base/java.util.stream.ReduceOps$4ReducingSink.accept(ReduceOps.java:220)
  at java.base/java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:992)
  at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
  at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
  at java.base/java.util.stream.ReduceOps$ReduceTask.doLeaf(ReduceOps.java:960)
  at java.base/java.util.stream.ReduceOps$ReduceTask.doLeaf(ReduceOps.java:934)
  at java.base/java.util.stream.AbstractTask.compute(AbstractTask.java:327)
  at java.base/java.util.concurrent.CountedCompleter.exec(CountedCompleter.java:754)
  at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373)
  at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1182)
  at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1655)
  at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1622)
  at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165)
```

**Desktop (please complete the following information):**
 - OS: WSL (Ubuntu 20.04)
 - Joern Version `2.0.103`
 - Java version:

**Comment by tuxology** [comment]
Wanted to leave an update here:

- The failure seems to be in cases where there are two secondary constructors defined. Eg:

```
class AClass {
    var x: String
    constructor(q: String) {
        this.x = q
    }
    constructor(p: String, r: Int) {
        this.x = p
    }
}
```

 - In such a case, we get a valid AST with one `(<init>, <init>)` call node attached to both constructor METHODs  - (not yet sure if this is relevant)
 - CFG for both constructors gets generated and seems correct. The incoming edges are also correct for METHOD_RETURN node.
 - In `ReachingDef` pass, the DataFlowSolver tries to find a solution given the CFG -  However, as it iterates over the worklist blocks in CFG it fails when computing the input set. Somehow the predecessor mapt doens't have the METHOD_RETURN node. 

Here is the expected DDG for the first constructor:
<img width="923" alt="image" src="https://github.com/joernio/joern/assets/3373857/70502151-1abf-4d11-8556-2d0e75cb867a">

Here is the complete AST for `AClass` type:
<img width="1212" alt="image" src="https://github.com/joernio/joern/assets/3373857/97e3fc8b-e24e-432d-9841-e022e8b4161a">

Any insights you may have will help me here @fabsx00 @ursachec👆

**Comment by ursachec** [comment]
@tuxology I'm not familiar enough with the dataflowengine to say anything on that front, but all in all it seems like you have a debugging issue on your hand – i.e. as a step forward, try to come up with the most minimal example which triggers the bug, put that in a test case, then use the debugger to find out why that crash is triggered and you'll probably have a solution, or at least a more specific question to ask

**Comment by tuxology** [comment]
@ursachec  I already have it in a test case. The example above is the most minimal example that I've added in the test case - the failing test is in my local branch on which I am using the IntelliJ debugger to debug it. Set breakpoints on the places where the bug triggers, step by step. I got some insights which I shared above and then followed up with investigating the AST and CFG on Joern for the CPG that was generated.

In addition in last few days, tried to test few things like creating separate call node for each constructor in AST. If you have any other insights, let me know.
